sweep/sweeper.go

```diff
@@ -1,234 +1,627 @@
+	// in case of a success, except for that a reorg could still happen.
+	if pendInput.ntfnRegCancel != nil {
+		pendInput.ntfnRegCancel()
+	}
+
+	// Inputs are no longer pending after result has been sent.
+	delete(s.pendingInputs, *outpoint)
+}
 
-	// With all the inputs in place, use each output's unique witness
-	// function to generate the final witness required for spending.
-	addWitness := func(idx int, tso Input) error {
-		witness, err := tso.BuildWitness(
-			s.cfg.Signer, sweepTx, hashCache, idx,
-		)
+// getInputLists goes through all pending inputs and constructs sweep lists,
+// each up to the configured maximum number of inputs. Negative yield inputs are
+// skipped. Transactions with an output below the dust limit are not published.
+// Those inputs remain pending and will be bundled with future inputs if
+// possible.
+func (s *UtxoSweeper) getInputLists(currentHeight int32,
+	satPerKW lnwallet.SatPerKWeight) ([][]Input, error) {
+
+	// TODO: Separate zero-attempt inputs from retries.
+
+	// Filter for inputs that need to be swept.
+	var sweepableInputs []Input
+	for _, input := range s.pendingInputs {
+		// Skip inputs that have a minimum publish height that is not
+		// yet reached.
+		if input.minPublishHeight > currentHeight {
+			continue
+		}
+
+		sweepableInputs = append(sweepableInputs, input.input)
+	}
+
+	return s.cfg.TxGenerator.Generate(sweepableInputs, satPerKW)
+}
+
+func (s *UtxoSweeper) sweep(inputs []Input,
+	satPerKW lnwallet.SatPerKWeight, currentHeight int32) error {
+
+	var err error
+
+	// Generate output script if no unused script available.
+	if s.currentOutputScript == nil {
+		s.currentOutputScript, err = s.cfg.GenSweepScript()
 		if err != nil {
 			return err
 		}
+	}
+
+	// Create sweep tx.
+	tx, err := s.cfg.TxGenerator.createSweepTx(inputs,
+		s.currentOutputScript,
+		uint32(currentHeight), satPerKW)
+	if err != nil {
+		return fmt.Errorf("cannot create sweep tx: %v", err)
+	}
 
-		sweepTx.TxIn[idx].Witness = witness
+	// Add tx before publication, so that we will always know that a
+	// spend by this tx is ours.
+	s.cfg.Store.NotifyPublishTx(tx)
 
-		return nil
+	// Publish sweep tx.
+	log.Debugf("Publishing sweep tx %v", tx.TxHash())
+	err = s.cfg.PublishTransaction(tx)
+
+	// Keep outputScript in case of an error, so that it can be
+	// reused for the next tx and causes no address inflation.
+	if err != nil {
+		log.Error(err)
+	} else {
+		s.currentOutputScript = nil
 	}
 
-	// Finally we'll attach a valid witness to each csv and cltv input
-	// within the sweeping transaction.
-	for i, input := range inputs {
-		if err := addWitness(i, input); err != nil {
-			return nil, err
+	// Reschedule sweep.
+	for _, input := range tx.TxIn {
+		pi := s.pendingInputs[input.PreviousOutPoint]
+
+		var nextAttemptDelta int32
+		switch err {
+		// Successful publish, but it may still not confirm.
+		// Retry later.
+		case nil:
+			nextAttemptDelta = 6
```

Comment:
magic constants?

Reply:
removed, because i simplified all constants to `1`

Reply:
okay, `1` was a bit too simple. changed back to exponential, but the constants didn't come back.